[RFC][RFT] uboot-sunxi: bump to 2021.07

mans0n opened the pull request:
Manually refreshed:
  093-sun6i-fix-PLL-LDO-voltselect.patch
  101-sun6i-support-console-on-UART2.patch

Deleted:
  210-sunxi-deactivate-binman.patch

As binman is re-activated, swig is now required to build U-Boot image [1].
Run-tested on PinePhone.

[1] https://lists.infradead.org/pipermail/openwrt-devel/2021-January/033605.html

mans0n committed:
uboot-sunxi: bump to 2021.07

Manually refreshed:
  093-sun6i-fix-PLL-LDO-voltselect.patch
  101-sun6i-support-console-on-UART2.patch

Deleted:
  210-sunxi-deactivate-binman.patch

As binman is re-activated, swig is now required to build U-Boot image [1].
Run-tested on PinePhone.

[1] https://lists.infradead.org/pipermail/openwrt-devel/2021-January/033605.html

Signed-off-by: Sungbo Eo <mans0n@gorani.run>
Tested-by: Aleksander Jan Bajkowski <olek2@wp.pl> [OPi One, OPi PC2]

mans0n commented:
Is it okay to require swig in OpenWrt repo? It is added to the buildbot dependency list, but I don't know whether the decision also applies here.

abajk commented:
My devices boot properly. Bootligs [1,2].

Tested-by: Aleksander Jan Bajkowski <olek2@wp.pl> # tested on Opi One, Opi PC2
[opione.log](https://github.com/openwrt/openwrt/files/7002306/opione.log)
[opipc2.log](https://github.com/openwrt/openwrt/files/7002307/opipc2.log)

dhewg commented:
This won't build anywhere if I understood correctly.
- swig is required. The package feeds has it. So this can easily be moved here with uboot-sunxi depending on swig/host
- the bigger issue is u-boot's pylibfdt, which seems to require the python headers. Moving python from the package feed here seems out of the question, I guess? Alternatively include/prereq-build.mk can check for it

1715173329 commented:
These patches may help:
https://github.com/openwrt/openwrt/blob/master/package/boot/uboot-rockchip/patches/001-scripts-remove-dependency-on-swig.patch
https://github.com/openwrt/openwrt/blob/master/package/boot/uboot-rockchip/patches/002-spl-remove-dtoc-of-pdata-generation.patch

mans0n commented:
@abajk Thanks for your test!

Now I found that python3-dev is already installed on my build machine. So I did some build-tests in Ubuntu 20.04 container.

2020.04: requires no extra packages
2021.07 in my PR: requires python3-dev, python3-distutils, python3-setuptools, swig
2021.07 with [001 patch](https://github.com/openwrt/openwrt/blob/master/package/boot/uboot-rockchip/patches/001-scripts-remove-dependency-on-swig.patch): requires python3-distutils and python3-setuptools, but the build fails due to binman
It seems [002 patch](https://github.com/openwrt/openwrt/blob/master/package/boot/uboot-rockchip/patches/002-spl-remove-dtoc-of-pdata-generation.patch) doesn't really help uboot-sunxi.

Current uboot-rockchip 2021.07: requires no extra package
I don't know why, but the build did not require python3-* or swig even after I removed aforementioned two patches.

Interestingly, people already have added python3-* and swig to Debian/Ubuntu prerequisites.
https://openwrt.org/docs/guide-developer/build-system/install-buildsystem

I'll try deactivating binman again if possible...

hauke commented:
I think more and more U-Boot builds will require binman and therefore need swig with python bindings. I would prefer to make this an official dependency for these U.Boot builds. If someone really can not install it he has to deactivate the U.Boot build.

mans0n committed:
uboot-sunxi: bump to 2022.01

Manually refreshed:
  230-disable-axp209-on-a13-olinuxino.diff

Deleted:
  250-sun8i-h3-zeropi-add-device-tree.patch
  251-sun8i-h3-zeropi-add-defconfig.patch
  253-sunxi-h5-add-support-for-nanopi-r1s-h5.patch

ZeroPi and NanoPi R1S H5 support was added to upstream U-Boot
with different code.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n commented:
I've finally spared some time to tie up these loose ends.

I assume we've decided to make swig a U-Boot build dependency, so I haven't investigated further on deactivating binman.
If we choose to use binman, then we also need python headers for pylibfdt and it seems inevitable (unless we use an external pylibfdt). Importing swig from the package feed was easy, but handling the python bindings was more difficult. After some trouble I've managed to add soft dependencies of python3-dev and python3-pkg-resources. I'm not sure about using `python3-config` to provide build options matching system python version.
And I've noticed that these prereqs are not checked again after one successful test, but target prereqs (in [x86](https://github.com/openwrt/openwrt/blob/628970a1959c0a0fbc0c3deb3d92945af63c2ddc/target/linux/x86/Makefile#L23-L28) and [uml](https://github.com/openwrt/openwrt/blob/628970a1959c0a0fbc0c3deb3d92945af63c2ddc/target/linux/uml/image/Makefile#L30-L33)) also show the same behavior.

While at it, I've also updated the U-Boot version to 2022.01, and it now complains about DTC provided by u-boot.mk. I thought it is not needed anymore so I just dropped it.

The local ZeroPi support patch was quite different from the upstream one, so there might be some regression.

I haven't dealt with U-Boot much before, so this definitely needs reviews. This is still a draft, and I'll do more build tests and tidy up the patchset before marking this as "ready".

mans0n commented:
Reduced the scope of change of DTC removal patch to uboot-sunxi. I'll open a new PR for the whole uboot targets after more tests.

hauke commented:
Why do you build swig in the tools folder and do not just add a build dependency for it? I think it is available in all relevant Linux distributions. Would we still need the python dev packages?

mans0n commented:
> Why do you build swig in the tools folder and do not just add a build dependency for it? I think it is available in all relevant Linux distributions.

You may be right, but swig is only required for U-boot build and I hesitate to add another hard dependency while we can just build it on demand.
Note that I placed swig in the packages directory, so it will be built only when required.

> Would we still need the python dev packages?

Yes, IIUC the dependency graph is:
```mermaid
graph TD;
  binman --> dtoc;
  dtoc --> pylibfdt;
  binman --> python3-pkg-resources;
  pylibfdt --> swig;
  pylibfdt --> python3-dev;
```

mans0n committed:
uboot-sunxi: bump to 2022.07

Manually refreshed:
  003-add-theobroma-a31-pangolin.patch
  100-sun6i-alternate-on-UART2.patch
  101-sun6i-support-console-on-UART2.patch
  252-sunxi-h3-add-support-for-nanopi-r1.patch

NanoPi R1 dts file was added to upstream U-Boot
with slightly different code.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n committed:
uboot-sunxi: force pylibfdt build

If we provide DTC environment variable then U-Boot also tries to use
system pylibfdt since upstream U-Boot commit 93b196532254 ("Makefile:
Only build dtc if needed"). Import a patch from uboot-mediatek to force
pylibfdt build even if the DTC env is provided.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n committed:
uboot-sunxi: add some prereq tests

Binman depends on libfdt and pkg_resources python modules, and
pylibfdt also requires swig and Python header. These are added
as local dependencies.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n committed:
uboot-sunxi: remove mach-sunxi Kconfig for pangolin

It is not required anymore.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n commented:
Rebased and updated to 2022.07.
And dropped swig in the tools directory and added it as a local dependency for now.

More thoughts:
* I've used `python3-config` in the python3-dev prereq test, but in some distros (like Ubuntu and Fedora) it is provided by the python3-dev package itself...
* Thanks to the pylibfdt patch in e0e74d8a2ced, I could safely drop my hack to mitigate the error. But still the pylibfdt patch does not spare us the trouble of installing python3-dev and python3-pkg-resources packages (even for uboot-mediatek). (It seems python3-pip provides them all in our buildbot setting) I believe those should be handled by explicit global/local dependencies.

hauke commented on package/boot/uboot-sunxi/patches/252-sunxi-h3-add-support-for-nanopi-r1.patch:
```diff
@@ -12,86 +12,54 @@ Signed-off-by: Jayantajit Gogoi <jayanta.gogoi525@gmail.com>
- 	sun8i-h3-orangepi-one.dtb \
---- /dev/null
+--- a/arch/arm/dts/sun8i-h3-nanopi-r1.dts
 +++ b/arch/arm/dts/sun8i-h3-nanopi-r1.dts
```
Is this patch needed at all any more? It looks like support for FriendlyARM NanoPi R1 was added upstream, our version is probably even older.

mans0n replied:
U-Boot imported the dts file from Linux kernel in u-boot/u-boot@2c597855aa17, but the defconfig for NanoPi R1 has not been added, so I left it in our patch.

And I first thought some nodes in the dts were overriden intentionally for our purpose (like reusing power button for restart), but I've found that the same dts was added for linux in e6d9f6fdffc6 and superseded by the upstreamed version, so now I think we can just drop the rest of the dts patch.

hauke commented on package/boot/uboot-sunxi/Makefile:
```diff
@@ -398,4 +398,17 @@ endef
+
+$(eval $(call TestHostCommand,python3-pkg-resources, \
+	Please install the Python3 pkg_resources module, \
+	$(STAGING_DIR_HOST)/bin/python3 -c 'import pkg_resources'))
```
I prefer to have this as a host dependency like your did.
Do you want to add this dependency only when uboot-sunxi is build?

Does it really work? It tried this some time ago and never got it really working, but I did it similarly to the thing in target/linux/x86/Makefile which I am pretty sure does not work.

mans0n replied:
> Do you want to add this dependency only when uboot-sunxi is build?

If possible, yes, I would add it to every uboot target that require binman build.

> Does it really work? It tried this some time ago and never got it really working, but I did it similarly to the thing in target/linux/x86/Makefile which I am pretty sure does not work.

Sort of.

>> I've noticed that these prereqs are not checked again after one successful test, but target prereqs (in x86 and uml) also show the same behavior.

The prereq check will run again after deleting and re-creating .config, so I think it's better than nothing.

KanjiMonster commented:
uboot sunxi has been bumped to 2023.04 in the mean time with 2fd049f5cdeebebfc91084ea316ac1fa2b84a369.

mans0n commented:
You mean 53c2f174ad79? Thanks for taking care of it anyway. :)

KanjiMonster commented:
Argh, yes, I did. Will I ever learn to properly handle the two different clipboards in linux?
